File: android-lib/src/main/java/com/kinvey/android/cache/QueryHelper.java
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */

package com.kinvey.android.cache;


import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Map;

import io.realm.DynamicRealmObject;
import io.realm.RealmQuery;

/**
 * Created by Prots on 2/1/16.
 */
public abstract class QueryHelper {

    public static RealmQuery<DynamicRealmObject>  prepareRealmQuery(RealmQuery<DynamicRealmObject> realmQuery, Map<String, Object> queryMap, boolean isIgnoreIn){
        for (Map.Entry<String, Object> entity : queryMap.entrySet()){
            String field = entity.getKey();
            Object params = entity.getValue();


            if (field.equalsIgnoreCase("$or")){
                realmQuery.beginGroup();
                if (params.getClass().isArray()){
                    Map<String, Object>[] components = (Map<String, Object>[])params;
                    if (components != null && components.length > 0) {
                        realmQuery.beginGroup();
                        prepareRealmQuery(realmQuery, components[0], isIgnoreIn);
                        realmQuery.endGroup();
                        for (int i = 1 ; i < components.length; i++) {
                            realmQuery.or();
                            realmQuery.beginGroup();
                            prepareRealmQuery(realmQuery, components[i], isIgnoreIn);
                            realmQuery.endGroup();
                        }
                    }
                }
                realmQuery.endGroup();
            } else  if (field.equalsIgnoreCase("$and")) {
                and(realmQuery, params, isIgnoreIn);
            } else if (params instanceof Map){
                for (Map.Entry<String, Object> paramMap : ((Map<String, Object>) params).entrySet()){
                    String operation = paramMap.getKey();
                    if (operation.equalsIgnoreCase("$in")){
                        if (!isIgnoreIn) {
                            in(realmQuery, field, paramMap.getValue());
                        } else {
                            realmQuery.beginGroup();
                            realmQuery.isNotEmpty("_id");
                            realmQuery.endGroup();
                        }
                    } else if (operation.equalsIgnoreCase("$nin")) {
                        realmQuery.beginGroup().not();
                        in(realmQuery, field, paramMap.getValue());
                        realmQuery.endGroup();
                    } else if (operation.equalsIgnoreCase("$gt")){
                        gt(realmQuery, field, paramMap.getValue());
                    } else if (operation.equalsIgnoreCase("$lt")){
                        lt(realmQuery, field, paramMap.getValue());
                    } else if (operation.equalsIgnoreCase("$gte")){
                        gte(realmQuery, field, paramMap.getValue());
                    } else if (operation.equalsIgnoreCase("$lte")){
                        lte(realmQuery, field, paramMap.getValue());
                    } else if (operation.equalsIgnoreCase("$ne")){
                        notEqualTo(realmQuery, field, paramMap.getValue());
                    } else {
                        throw new UnsupportedOperationException("this query is not supported by cache");
                    }

                }
            } else {
                equalTo(realmQuery, field, params);
            }
        }
        return realmQuery;
    }

    public static RealmQuery<DynamicRealmObject>  prepareRealmQuery(RealmQuery<DynamicRealmObject> realmQuery, Map<String, Object> queryMap){
        return prepareRealmQuery(realmQuery, queryMap, false);
    }

    private static void and(RealmQuery query, Object params, boolean isIgnoreIn){
        query.beginGroup();
        if (params.getClass().isArray()) {
            Map<String, Object>[] components = (Map<String, Object>[]) params;
            if (components != null && components.length > 0) {
                query.beginGroup();
                prepareRealmQuery(query, components[0], isIgnoreIn);
                query.endGroup();
                for (int i = 1; i < components.length; i++) {
                    query.beginGroup();
                    prepareRealmQuery(query, components[i], isIgnoreIn);
                    query.endGroup();
                }
            }
        }
        query.endGroup();
    }

    private static void in(RealmQuery query, String field, Object params){
        if (params.getClass().isArray()) {
            Object[] operatorParams = (Object[]) params;
            query.beginGroup();
            if (operatorParams != null && operatorParams.length > 0) {
                equalTo(query, field, operatorParams[0]);
                for (int i = 1; i < operatorParams.length; i++) {
                    query.or();
                    equalTo(query, field, operatorParams[i]);
                }
            }
            query.endGroup();
        }
    }

    private static void gt(RealmQuery query, String field, Object param){
        try {
            if (param instanceof Number) {

                Method m = query.getClass().getMethod("greaterThan", String.class,
                        (Class) param.getClass().getDeclaredField("TYPE").get(param));
                m.invoke(query, field, param);
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }
    }

    private static void gte(RealmQuery query, String field, Object param){
        try {
            if (param instanceof Number) {

                Method m = query.getClass().getMethod("greaterThanOrEqualTo", String.class,
                        (Class) param.getClass().getDeclaredField("TYPE").get(param));
                m.invoke(query, field, param);
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }
    }

    private static void lte(RealmQuery query, String field, Object param){
        try {
            if (param instanceof Number) {

                Method m = query.getClass().getMethod("lessThanOrEqualTo", String.class,
                        (Class) param.getClass().getDeclaredField("TYPE").get(param));
                m.invoke(query, field, param);
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }
    }

    private static void lt(RealmQuery query, String field, Object param){

        try {
            if (param instanceof Number){
                Method m = query.getClass().getMethod("lessThan", String.class,
                        (Class)param.getClass().getDeclaredField("TYPE").get(param));
                m.invoke(query, field, param);
            }

        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }
    }

    private static void equalTo(RealmQuery query, String field, Object param){

        try {
            Method m = query.getClass().getMethod("equalTo", String.class,
                    param.getClass());
            m.invoke(query, field, param);
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

    }

    private static void notEqualTo(RealmQuery query, String field, Object param){
        try {
            Method m = query.getClass().getMethod("notEqualTo", String.class, param.getClass());
            m.invoke(query, field, param);
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

    }

}

File: android-lib/src/main/java/com/kinvey/android/cache/QueryHelper.kt
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */

package com.kinvey.android.cache

import io.realm.DynamicRealmObject
import io.realm.RealmQuery
import java.lang.reflect.InvocationTargetException
import java.util.*

/**
 * Created by Prots on 2/1/16.
 */
object QueryHelper {
    @JvmStatic
    fun prepareRealmQuery(realmQuery: RealmQuery<DynamicRealmObject>,
                          queryMap: Map<String, Any>?, isIgnoreIn: Boolean)
            : RealmQuery<DynamicRealmObject> {
        for ((field, params) in queryMap ?: mapOf()) {
            when {
                field.equals("\$or", ignoreCase = true) -> {
                    realmQuery.beginGroup()
                    if (params.javaClass.isArray) {
                        val components = params as Array<Map<String, Any>>
                        if (components != null && components.size > 0) {
                            realmQuery.beginGroup()
                            prepareRealmQuery(realmQuery, components[0], isIgnoreIn)
                            realmQuery.endGroup()
                            for (i in 1 until components.size) {
                                realmQuery.or()
                                realmQuery.beginGroup()
                                prepareRealmQuery(realmQuery, components[i], isIgnoreIn)
                                realmQuery.endGroup()
                            }
                        }
                    }
                    realmQuery.endGroup()
                }
                field.equals("\$and", ignoreCase = true) -> and(realmQuery, params, isIgnoreIn)
                params is Map<*, *> -> for ((operation, value) in params as Map<String, Any>) {
                    if (operation.equals("\$in", ignoreCase = true)) {
                        if (!isIgnoreIn) {
                            `in`(realmQuery, field, value)
                        } else {
                            realmQuery.beginGroup()
                            realmQuery.isNotEmpty("_id")
                            realmQuery.endGroup()
                        }
                    } else if (operation.equals("\$nin", ignoreCase = true)) {
                        realmQuery.beginGroup().not()
                        `in`(realmQuery, field, value)
                        realmQuery.endGroup()
                    } else if (operation.equals("\$gt", ignoreCase = true)) {
                        gt(realmQuery, field, value)
                    } else if (operation.equals("\$lt", ignoreCase = true)) {
                        lt(realmQuery, field, value)
                    } else if (operation.equals("\$gte", ignoreCase = true)) {
                        gte(realmQuery, field, value)
                    } else if (operation.equals("\$lte", ignoreCase = true)) {
                        lte(realmQuery, field, value)
                    } else if (operation.equals("\$ne", ignoreCase = true)) {
                        notEqualTo(realmQuery, field, value)
                    } else if (operation.equals("\$not", ignoreCase = true)) {
                        val newMap: MutableMap<String, Any> = LinkedHashMap()
                        newMap[field] = value as Map<String?, Any?>
                        realmQuery.not()
                        prepareRealmQuery(realmQuery, newMap, isIgnoreIn)
                    } else {
                        throw UnsupportedOperationException("this query is not supported by cache")
                    }
                }
                else -> equalTo(realmQuery, field, params)
            }
        }
        return realmQuery
    }

    @JvmStatic
    fun prepareRealmQuery(realmQuery: RealmQuery<DynamicRealmObject>, queryMap: Map<String, Any>?): RealmQuery<DynamicRealmObject> {
        return prepareRealmQuery(realmQuery, queryMap, false)
    }

    private fun and(query: RealmQuery<DynamicRealmObject>, params: Any, isIgnoreIn: Boolean) {
        query.beginGroup()
        if (params.javaClass.isArray) {
            val components = params as Array<Map<String, Any>>
            if (!components.isNullOrEmpty()) {
                query.beginGroup()
                prepareRealmQuery(query, components[0], isIgnoreIn)
                query.endGroup()
                for (i in 1 until components.size) {
                    query.beginGroup()
                    prepareRealmQuery(query, components[i], isIgnoreIn)
                    query.endGroup()
                }
            }
        }
        query.endGroup()
    }

    private fun `in`(query: RealmQuery<*>, field: String, params: Any) {
        if (params.javaClass.isArray) {
            val operatorParams = params as Array<Any>
            query.beginGroup()
            if (operatorParams != null && operatorParams.size > 0) {
                equalTo(query, field, operatorParams[0])
                for (i in 1 until operatorParams.size) {
                    query.or()
                    equalTo(query, field, operatorParams[i])
                }
            }
            query.endGroup()
        }
    }

    private fun gt(query: RealmQuery<*>, field: String, param: Any) {
        try {
            if (param is Number) {
                val m = query.javaClass.getMethod("greaterThan", String::class.java,
                        param.javaClass.getDeclaredField("TYPE").get(param) as Class<*>)
                m.invoke(query, field, param)
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
    }

    private fun gte(query: RealmQuery<*>, field: String, param: Any) {
        try {
            if (param is Number) {
                val m = query.javaClass.getMethod("greaterThanOrEqualTo", String::class.java,
                        param.javaClass.getDeclaredField("TYPE").get(param) as Class<*>)
                m.invoke(query, field, param)
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
    }

    private fun lte(query: RealmQuery<*>, field: String, param: Any) {
        try {
            if (param is Number) {
                val m = query.javaClass.getMethod("lessThanOrEqualTo", String::class.java,
                        param.javaClass.getDeclaredField("TYPE").get(param) as Class<*>)
                m.invoke(query, field, param)
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
    }

    private fun lt(query: RealmQuery<*>, field: String, param: Any) {
        try {
            if (param is Number) {
                val m = query.javaClass.getMethod("lessThan", String::class.java,
                        param.javaClass.getDeclaredField("TYPE").get(param) as Class<*>)
                m.invoke(query, field, param)
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
    }

    private fun equalTo(query: RealmQuery<*>, field: String, param: Any) {
        try {
            val m = query.javaClass.getMethod("equalTo", String::class.java,
                    param.javaClass)
            m.invoke(query, field, param)
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }

    private fun notEqualTo(query: RealmQuery<*>, field: String, param: Any) {
        try {
            val m = query.javaClass.getMethod("notEqualTo", String::class.java, param.javaClass)
            m.invoke(query, field, param)
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }
}